# Alpha-Index, Maximin, and Leximin Rules for Decisions Under Uncertainty

## Alpha-Index

Originally proposed by [Hurwicz](https://en.wikipedia.org/wiki/Leonid_Hurwicz), the $\alpha$-index framework for decision-making considers both the best and worst possible outcomes of each alternative. The rule dictates that one chooses an alternative according to one's degree of optimism (or pessimism). So:
 
$$
AlphaIndex=\alpha \bullet max(a_i)+(1-\alpha)\bullet min(a_i)
$$

---

Assuming:

- A decision maker's degree of optimism can be represented by a real number, $\alpha$ between $0$ and $1$.
- $\alpha$ can rationally assumed to be fixed throughout the evaluation of alternatives, $a_i$.

---

Therefore, we have: 

$$
a_i \succ a_j \Leftrightarrow \alpha \bullet max(a_i)+(1-\alpha)\bullet min(a_i) > \alpha \bullet max(a_j)+(1-\alpha)\bullet min(a_j) 
$$




As an example consider the following decision matrix:

$$
\begin{matrix}
    a_1 & 14 & 12 & 27 & 10 & 36 & 100\\
    \hline
    a_2 & 30 & 19 & 66 & 43 & 47 & 77
\end{matrix}\qquad
$$

We can use numpy to help us make decisions in this scenario, given an $\alpha$-level of $0.7$:

In [1]:
import numpy as np
A = np.array([14, 12, 27, 10, 36, 100])
B = np.array([30, 19, 66, 43, 47, 77])

alpha = 0.7
l = alpha*np.max(A)+(1-alpha)*np.min(A)
h = alpha*np.max(B)+(1-alpha)*np.min(B)

print(l,h)

73.0 59.6


Since, $73>59.6$, it follows that alternative $a_1$ is better than alternative $a_2$.

---

## Maximin

The so-called maximin principle places all weight in the decision calculus on the worst case scenario. According to this view, one should choose so as to maximize the minimal value between possible actions. This view can formalized as follows, given that $min(a_i)$ is the minimal value obtainable with choice $a_i$. Then we have: 

$$
a_i \succeq a_j \Leftrightarrow min(a_i) \geq min(a_j)
$$


As an example, consider the following decision matrix:

$$
\begin{matrix}
    a_1 & 4 & 12 & 11 & 0\\
    \hline
    a_2 & 6 & -4 & 66 & 143\\
    \hline
    a_3 & 5 & 7 & 1 & 6
\end{matrix}\qquad
$$

The maximin rule demands that one choose $a_3$, since the worst case scenario for $a_3$ is $1$, which is better than the worst-case scenarios for $a_1$ and $a_2$. We can confirm this with Python: 

In [2]:
import numpy as np
m = np.array([[4, 12, 11, 0], [6, -4, 66, 143], [5, 7, 1, 6]])

def maximin(m):
    return m.min(axis=1)

In [3]:
maximin(m)

array([ 0, -4,  1])

---

## Leximin

One alternative to decision-making in this context is the so-called leximin rule. Like the maximin rule, it holds that decision makers should maximize the minimal value obtainable with each act. However, if the worst outcomes available are equally bad, one should choose the alternative in which the second worst outcome is as good as possible. Unless all possible outcomes are exactly parallel (with respect to their value), the leximin rule will, at some point, find an act that is better than its alternatives. This leads to the following formalization:

$$
a_i \succ a_j \Leftrightarrow \exists n : min^n(a_i) > min^n(a_j) \land min^m(a_i) = min^m(a_j) \, \forall  (m < n)
$$

As an example, consider the following decision matrix:

$$
\begin{matrix}
    a_1 & 4 & 12 & 11 & 6\\
    \hline
    a_2 & 6 & 3 & 66 & 143\\
    \hline
    a_3 & 5 & 5 & 5 & 6
\end{matrix}\qquad
$$

In this scenario, the leximnin rule would recommend $a_3$. We can confirm this with Python:  

In [6]:
import numpy as np
from collections import Counter
mask = 100000000
m = np.array([[4, 12, 11, 6], [6, 3, 66, 143], [5, 5, 5, 6]])


def leximin(m):
    if np.all(m == m[0,:]):
        return m.min(axis=1)
    elif np.all(m != m[0,:]):
        i = m.min(axis= 1)
        j = np.unique(i).size
        if j == len(m): 
            return i
    else:
        i = m.min(axis= 1)
        j = np.unique(i).size
        while j != len(m):
            clone=[item for item, count in Counter(i).items() if count > 1]
            matrix_copy = np.ndarray.flatten(m)
            index=np.where(np.in1d(matrix_copy,clone))
            matrix_copy[index] = mask 
            m = matrix_copy.reshape(m.shape[0],m.shape[1])    
            i = m.min(axis= 1)
            j = np.unique(i).size
            
        return i

In [7]:
leximin(m)

array([4, 3, 5])